src/mcp/tools/rules/validate_rules.ts

```diff
@@ -0,0 +1,137 @@
+        ) {
+          const startColumnOnLine = sourcePosition.column - 1; // 0-based start column for carets
+          const errorTokenLength = sourcePosition.endOffset - sourcePosition.currentOffset;
+
+          // Ensure startColumnOnLine is not negative and errorTokenLength is positive
+          // Also, ensure startColumnOnLine is within the line length (or reasonably close for safety)
+          if (
+            startColumnOnLine >= 0 &&
+            errorTokenLength > 0 &&
+            startColumnOnLine <= errorLine.length
+          ) {
+            const padding = " ".repeat(startColumnOnLine);
+            const carets = "^".repeat(errorTokenLength);
+            issueString += `\n${padding}${carets}\n\`\`\``;
+          }
+        }
+      }
+    }
+    formattedOutput.push(issueString);
+  }
+  return formattedOutput.join("\n\n"); // Separate multiple issues with a blank line
+}
+
+export function validateRulesTool(productName: string) {
+  return tool(
+    {
+      name: "validate_rules",
+      description: `Checks the provided ${productName} Rules source for syntax and validation errors. Provide EITHER the source code to validate OR a path to a source file.`,
+      inputSchema: z.object({
+        source: z
+          .string()
+          .optional()
+          .describe("the rules source code to check. provide either this OR a path"),
+        source_file: z
+          .string()
+          .optional()
+          .describe(
+            "a file path, relative to the project root, to a file containing the rules source you want to validate. provide this OR source, not both",
+          ),
+      }),
+      annotations: {
+        title: `Validate ${productName} Rules`,
+        readOnlyHint: true,
+      },
+      _meta: {
+        requiresProject: true,
+        requiresAuth: true,
+      },
+    },
+    async ({ source, source_file }, { projectId, config }) => {
+      if (source && source_file) {
+        return mcpError("Must supply `source` or `source_file`, not both.");
+      }
+
+      let rulesSourceContent: string;
+      if (source_file) {
+        try {
+          rulesSourceContent = readFileSync(config.readProjectFile(source_file)).toString();
+        } catch (e: any) {
+          return mcpError(`Failed to read source_file '${source_file}': ${e.message}`);
+        }
+      } else if (source) {
+        rulesSourceContent = source;
+      } else {
+        // If neither source nor source_file is provided, default to empty string for validation.
+        // testRuleset might still return general errors not tied to specific lines.
+        rulesSourceContent = "";
+      }
+
+      const result = await testRuleset(projectId!, [
+        // The name "firestore.rules" is a convention for testRuleset,
+        // actual fileName from issues will be used in formatting.
+        { name: "firestore.rules", content: rulesSourceContent },
```

Comment:
If this name don't matter, can we do something generic like `test.rules`?

Right now, it looks like a bug if this is used to test `storage.rules`.